### ISSUE TITLE
series breaks on draft articles

### ISSUE THREAD

**Issue by mpounsett** [pelican-plugins/series#31]
This is similar to #18, although the specifics I'm seeing are slightly different than what was described there.

I've got a draft article with my first series in it.. and just added series support to my theme.  My pelicanconf.py defines DEFAULT_METADATA of `"status": "published"` for development purposes, and publishconf.py sets that default to 'draft'.  This is the setup I use so that the devserver will show me draft articles as they will appear once published.  My uses of `pelican-series` worked fine in this setup until I tried to do a producton publish of a new article, while the unfinished series article is still in drafts.

When Pelican's `make publish` target hits that draft article, it fails.  The first encountered error is a reference to `article.series.previous.url`:

```
UndefinedError: 'str object' has no attribute 'previous'
```
The error comes from this bit of template in my `article.html`:
``` jinja2
{% if article.series.previous is not none %}
    <a class="series-footer text-muted"
       href="{{ SITEURL }}/{{ article.series.previous.url }}">
        <i class="fa fa-angles-left"></i>
        Previous Article
    </a>
{% else %}
    <i class="fa fa-angles-left"></i>
    Previous Article
{% endif %}
```

That block is itself wrapped in an if to check whether the article is part of a series:
``` jinja2
{% if SHOW_SERIES and article.series is defined %}
```

The draft in question is the first and only article in the series, so `article.series.next` and `article.series.previous` should both be `None`.  I've done some strategic commenting-out of bits of the template and inserting values as comments to determine that `article.series.previous` actually references `article.series.name` in a draft, rather than `article.series.all_previous[-1]`, as it does when the article is published.

Oddly, `article.series.index` and `article.series.name` both seem to be empty string when the document is a draft, but work as expected when the document is published.

Relevant software version info from my virtualenv:

``` console
% python --version
Python 3.10.9

% pip freeze
blinker==1.5
commonmark==0.9.1
docutils==0.19
feedgenerator==2.0.0
Jinja2==3.1.2
MarkupSafe==2.1.1
pelican==4.8.0
pelican-i18n-subsites==0.9.0
pelican-series==2.1.0
pelican-youtube==0.2.1
Pygments==2.13.0
python-dateutil==2.8.2
pytz==2022.7
rich==13.0.0
six==1.16.0
smartypants==2.0.1
typogrify==2.0.7
Unidecode==1.3.6
```

**Comment by lgiordani** [comment]
@mpounsett I'm so sorry I ignored this error for so long. Would you be able to provide a GH repo with a minimal example of the situation? This way it will be very easy to debug. Thanks, and again so sorry for the delay.

**Comment by mpounsett** [comment]
Hey @lgiordani ... now it's my turn to have been ignoring this. :) I set up a workaround (series in a different drafts folder) and never got around to returning to the issue.

I did do a little more investigating at some point, and in 2.1.0 the problem seemed to be [here](https://github.com/pelican-plugins/series/blob/147e2617d636da0e8fa27006217020f8c576eb36/pelican/plugins/series/series.py#L26-L35):

```python
    for article in generator.articles:

        if "series" in article.metadata:
            article_entry = {
                "index": int(article.metadata.get("series_index", 0)),
                "date": article.metadata["date"],
                "content": article,
            }

            series[article.metadata["series"]].append(article_entry)
```

It looked to me like that was only collecting (published) articles, and not drafts, to update the article entry with the series metadata.  And then, if I remember correctly, somewhere down the line that meant that the `article.series` attribute in the template gets left as a string and doesn't get replaced with the series object (which has the `next` and `previous` attributes).

I should stress this was a while ago and I'm half-remembering my reasoning.  However, this patch fixed the issue in 2.10:

```diff
--- series.py.orig	2026-02-10 10:48:18
+++ series.py	2026-02-10 10:52:45
@@ -14,7 +14,9 @@

 from pelican import signals

+from itertools import chain

+
 def aggregate_series(generator):
     generator.context["series"] = {}

@@ -23,7 +25,9 @@
     # This cycles through all articles in the given generator
     # and collects the 'series' metadata, if present.
     # The 'series_index' metadata is also stored, if specified
-    for article in generator.articles:
+    for article in chain(
+        generator.articles, generator.drafts, generator.drafts_translations
+    ):

         if "series" in article.metadata:
             article_entry = {
```

3.1.0 still exhibits the problem.  The code has changed a fair bit, but it looks like the same structure still exists [here](https://github.com/pelican-plugins/series/blob/f6428bea25285bf8043d17b8c3189e5ec8935cc9/pelican/plugins/series/series.py#L62).   The same change doesn't fix it, but the pelican API seems to have moved on, so it's likely the patch is no longer referencing all the right generator attributes or something.   The 2.1.0 fix might give you ideas though?

I'll see if I can put together a MVE ... how minimal does it need to be?  Is it good enough if I find a standard pelican theme and a single draft that triggers the issue?

**Comment by mpounsett** [comment]
In doing the writeup below, I realized there was a key piece of info missing from my original problem report (that I didn't realize at the time was key) which is ... this breaks when articles marked as `status: draft` get **published** into the `output/drafts/` directory during a `make publish`.   The problem being that pelican-series ignores drafts when it's building the series-specific article metadata.

Anyhow... as as I was cleaning up from the temporary patching I'd done for my previous comment, I realized I'd implemented my test patch on 3.1.0 in the wrong place—not paying enough attention, I put it on the wrong loop.   So I did more careful digging into the Pelican classes that series touches, and re-implemented my suggested fix for 3.1.0.  I'm still not super familiar with the Pelican internals so please do not assume my suggestion below is sufficient or complete ... but it does seem to fix the problem for my templates.

```diff
--- series.py.orig	2026-02-10 12:31:36
+++ series.py	2026-02-10 12:40:08
@@ -12,6 +12,8 @@

 from pelican import signals

+from itertools import chain
+
 ordered_articles_all = {}
 ordered_pages_all = {}

@@ -58,11 +60,25 @@
     # and collects the 'series' metadata, if present.
     # The 'series_index' metadata is also stored, if specified
     translations = defaultdict(dict)
+    base_articles = list(
+        chain(
+            generator.articles,
+            generator.translations,
+            generator.drafts,
+            generator.drafts_translations,
+            generator.hidden_articles,
+            generator.hidden_translations,
+        )
+    )
     more = []
-    for article in generator.articles:
-        for trans in article.translations:
-            more.append(trans)
-    for article in generator.articles + more:
+    for article in base_articles:
+        more.extend(article.translations)
+    seen = set()
+    for article in chain(base_articles, more):
+        key = id(article)
+        if key in seen:
+            continue
+        seen.add(key)
         if "series" in article.metadata:
             article_entry = {
                 "index": int(article.metadata.get("series_index", 0)),
```

Here's an explanation of my changes, and why I made them.  Hopefully this helps you figure out where I've either missed things or have gone overboard. 😁 

- `base_articles`:  
   All six of those attributes on the `ArticlesGenerator` class can contain articles, so they're all candidates for having series data assigned to them.   The bug I'm experiencing would also have affected anyone using series on translations or hidden articles.
  I'm collapsing the generator returned by `itertools.chain()` into a list, rather than leaving `base_articles` as a generator, because it needs to be iterated over more than once in the following lines.
- `more`:  
  I'm doing this _after_ generating `base_articles` because everything in `base_articles` has a `translations` attribute that might contain more articles that need to be examined.   My use of `list.extend()` to do the assignment is functionally identical to what was there before, but avoids running a loop over every single translation.
- `seen`:  
  This might be overkill, but while I was working on this the idea occurred to me that individual translations might show up in more than one of the lists.  e.g.  maybe the same document appears in generator.translations and generators.articles[x].translations.  I don't know if this is actually possible.  Assuming it is possible, the use of `seen` will do some deduplication to avoid touching an article more than once.   If articles can't appear in multiple lists simultaneously, then this is a waste of effort, and should be removed.

`ArticlesGenerator` has some other attributes that look like they might contain articles, but I didn't dig far enough to find out for sure:
- `related_posts` seems likely, but I wasn't sure if it made sense to process those for series data.  _Probably_ anything in `related_posts` is also in one of the other article-collecting attributes?
- `period_archives` and `categories` don't seem super likely, but because I don't know for sure what all can possibly appear in those lists I'm mentioning them to be safe

If you think this does look like a complete fix then I can whip up a PR.  If more investigation into the bug is needed, let me know and I can still do that MVE, if you want.  For the moment I'm not going to put effort into that... but let me know if you need it and I can put something together.

**Comment by lgiordani** [comment]
Hi @mpounsett, no worries for the delay. Thanks for the long report, very useful. I'm going to have a look as soon as possible, I promise I won't ignore it! ✋

**Comment by lgiordani** [comment]
Hi @mpounsett. Can you please double check that the PR above solves your problem? I think the solution was very straightforward, I don't know why it took me so long!

**Comment by mpounsett** [comment]
> I think the solution was very straightforward, I don't know why it took me so long!

Probably because when I wrote my original issue report I didn't know much about how the internals were organized, and so there wasn't much to go on for finding the source of the issue.  😁 


I can see just from reviewing the code that the PR will solve my immediate problem.  

I had two "is this even a problem?" things in my fix which you didn't cover... so just to check, are these statements both true?

1. Articles in `generator.articles[i].translations` always appear in `generator.translations`. (so processing `generator.articles[i].translations` is unnecessary)
2. Articles never appear in more than one of the processed lists. (so deduplication is not required)

If both are true, then this fixes the problem completely (not just my specific issue) as far as I can see.   If 2 is false, it's still possible to say you don't care about potentially double-processing articles, and so deduplication is unnecessary.

**Comment by lgiordani** [comment]
Yes, apologies for not replying directly to those. I'm not a core Pelican developer, so I had to do some research 😹
As far as I understand this is what happens:

1. The function `process_translations` in Pelican [code](https://github.com/getpelican/pelican/blob/8b4d7475ea9ee7f998f482ce6fc1f5267e12a73d/pelican/utils.py#L652) returns a disjoint list: originals (`generator.articles`) and translations (`generator.translations`). Then `article.translations` is set to point to the siblings in that list. Basically, iterating on `generator.translations` is equivalent to going through `article.translations`.
2. The function `generate_context` [code](https://github.com/getpelican/pelican/blob/8b4d7475ea9ee7f998f482ce6fc1f5267e12a73d/pelican/generators.py#L661) sorts each article into one of the lists based on its `status`, then `_process` (inside the function) splits originals and translations. In the end we have six disjoint lists: articles, drafts, translations, and the hidden version of all three.

So, I _believe_ both your statements are true, and this fix is enough. Does it make sense?

**Comment by lgiordani** [comment]
Hi @mpounsett! Sorry for the nudge, I just wanted to check if you think the answers above are satisfactory enough, so that we can merge the PR. Thanks!

Leo

**Comment by mpounsett** [comment]
Yeah, makes sense to me.  I only raised the questions to be sure they'd been considered, and they clearly have been.  So yes, I think this looks great.
